nemo/core/classes/module.py

```diff
@@ -58,6 +58,21 @@ def freeze(self) -> None:
         for param in self.parameters():
             param.requires_grad = False
 
+        # By default, if you run a batch norm module with
+        # training=False (calling self.eval() below makes
+        # training=False for a module and all its descendents),
+        # pytorch will still update batch norm statistics. This can
+        # cause a denial of service attack, because an adversary can
+        # provide an input that causes the batch norm statistics to
+        # get set to NaN or infinity. Setting track_running_stats to
+        # False prevents updates of the batch norm statistics.
+        for sub_module in self.modules():
+            if isinstance(sub_module,
+                          (torch.nn.BatchNorm1d, torch.nn.LazyBatchNorm1d, torch.nn.BatchNorm2d,
+                           torch.nn.LazyBatchNorm2d, torch.nn.BatchNorm3d, torch.nn.LazyBatchNorm3d,
+                           torch.nn.SyncBatchNorm)):
+                sub_module.track_running_stats = False
```

Comment:
I might be wrong. Isn't track running stats necessary for grad accumulation calculation?

Reply:
This in the freeze() method. freeze() sets requires_grad=False for all parameters. and sets self.eval(). Why would we need gradients in this situation?

Anyway, to answer your question, no you don't need to track running stats to calculate the gradient through batch norm. (Might be useful if you freeze a discriminator, but not a generator? IDK.)

Reply:
Okay, sounds good. I will remove this change.

BTW, for the sake of good epistemics, I reran the test I did in the first comment with track_running_stats = False and with track_running_stats = True. The WER score does not change at all, for any dataset, up to 8 decimal places after the decimal point, for either the casting mode or AMP.